### ISSUE TITLE
Question: Log transport

### ISSUE THREAD

**Issue by visiongeist** [Microsoft/ApplicationInsights-node.js#498]
Hi,

since loggers like winston are instrumented, does this mean all logs are transported to Azure? I haven't found a way to access them in the portal.

Thank you

**Comment by markwolff** [comment]
Hi @visiongeist It should be transporting them to Azure. Which version of winston are you using? We support winston 2.x and 3.x. 3.x support was added in 1.2.0 of this library. If this all sounds fine to you and it is still not working can you post a repro of how you are setting up winston and logging with it?

**Comment by visiongeist** [comment]
Hi,

Thank you for your response!
 
I'm using 2.4.1 and I'm not seeing any logs under Search even though application map works!

`
const appInsights = require('applicationinsights');
appInsights.setup().start(); // using ENV variable

const winston = require('winston');
winston.log('info', `Start server ${process.env.npm_package_version}`)
`

**Comment by markwolff** [comment]
@visiongeist I'm unable to repro this -- it seems to be showing up for me. Are you requiring in winston somewhere else first in your app (e.g. in a different file)? For the auto transporting to work, App Insights needs to be the first thing you require (or at least be before everything that you want it to monitor).

```js
const winston = require('winston');
setInterval( () => winston.log('info', `Starting server (my winston logger)`), 10000);
```

![image](https://user-images.githubusercontent.com/5515583/55267748-6ad1d880-5241-11e9-8bca-b519be823a6a.png)

**Comment by aparralorenzo** [comment]
Hi, 

I have same issue. 
I can see the dependencies in the portal but not the traces from winston:
![socket](https://user-images.githubusercontent.com/3873409/55549734-dbac4100-56d6-11e9-839a-bc6b6d897c3f.jpg)

I can see Winston logs in the console and the file
![socketCosonle](https://user-images.githubusercontent.com/3873409/55549733-dbac4100-56d6-11e9-80bd-3b85da973a21.jpg)

Our first require is applicationinsights
   "applicationinsights": "1.3.0",
    "body-parser": "1.18.3",
    "compression": "1.7.3",
    "cors": "2.8.5",
    "express": "4.16.4",
    "ioredis": "4.6.2",
    "socket.io": "2.2.0",
    "socket.io-redis": "5.2.0",
    "spdy": "4.0.0",
    "uuid": "3.3.2",
    "winston": "3.2.1"

index.js
```javascript
const appInsights = require('applicationinsights');
const spdy = require('spdy');
const fs = require('fs');
const path = require('path');
const app = require('./app');

const config = require('./config');

appInsights
  .setup(config.INSTRUMENTATION_KEY)
  .setAutoDependencyCorrelation(true)
  .setAutoCollectRequests(true)
  .setAutoCollectPerformance(true)
  .setAutoCollectExceptions(true)
  .setAutoCollectDependencies(true)
  .setAutoCollectConsole(true)
  .setUseDiskRetryCaching(true);
appInsights.defaultClient.context.tags[appInsights.defaultClient.context.keys.cloudRole] =
  'sockets-server';
appInsights.start();

const { SERVER_KEY_PATH, SERVER_CRT_PATH, SERVER_INTERM_PATH } = process.env;
let serverOptions = {};
if (SERVER_KEY_PATH || SERVER_CRT_PATH || SERVER_INTERM_PATH) {
  serverOptions = {
    key: fs.readFileSync(SERVER_KEY_PATH),
    cert: fs.readFileSync(SERVER_CRT_PATH),
    ca: fs.readFileSync(SERVER_INTERM_PATH),
  };
} else {
  serverOptions = {
    key: fs.readFileSync(path.join(__dirname, '../server.key')),
    cert: fs.readFileSync(path.join(__dirname, '../server.crt')),
  };
}

const server = spdy.createServer(serverOptions, app).listen(app.get('port'), (error) => {
  if (error) {
    console.error(error); 
  }
  console.log(`Server listening on port ${app.get('port')}`); 
});

require('./sockets/startSocketsServer')(server);
```

loggers.js
```javascript
const { createLogger, format } = require('winston');

const { combine, timestamp, label, printf } = format;

const prettifyMessage = (message) => JSON.stringify(message, null, 2);
const myFormat = printf(
  (info) =>
    `${info.timestamp} - ${info.id} | [${info.label}] (${info.level}): ${prettifyMessage(
      info.message,
    )}`,
);

const myLogger = createLogger({
  level: 'info',
  format: combine(label({ label: 'myLogger ' }), timestamp(), myFormat),
});

const otherLogger = createLogger({
  level: 'info',
  format: combine(label({ label: 'otherLogger ' }), timestamp(), myFormat),
});

module.exports = {
  myLogger, 
  otherLogger ,
};

```
transports.js
```javascript
const path = require('path');
const fs = require('fs');
const { transports } = require('winston');

const logsPath = path.join(process.cwd(), 'logs');

if (!fs.existsSync(logsPath)) {
  fs.mkdirSync(logsPath);
}

const infoLogFilename = path.join(logsPath, 'info.log');
const errorsLogFilename = path.join(logsPath, 'errors.log');

const fileTransports = [
  new transports.File({
    filename: infoLogFilename,
    maxsize: 5000000,
    maxFiles: 20,
  }),
  new transports.File({
    filename: errorsLogFilename,
    handleExceptions: true,
    level: 'error',
    maxsize: 5000000,
    maxFiles: 20,
  }),
];

const consoleTransports = [new transports.Console()];

const productionTransports = [...fileTransports];
const developmentTransports = [...consoleTransports, ...fileTransports];

module.exports = { productionTransports, developmentTransports };

```
initlogger.js

```javascript
const { loggers } = require('winston');
const { productionTransports, developmentTransports } = require('../utils/logger/transports');
const { myLogger, otherLogger } = require('../utils/logger/loggers');

const addTransports = () => {
  if (process.env.NODE_ENV !== 'production') {
    loggers.options.transports = developmentTransports;
  } else {
    loggers.options.transports = productionTransports;
  }
};

const addLoggers = () => {
  loggers.add('MY', myLogger);
  loggers.add('OTHER', otherLogger );
};

module.exports = function initLogger() {
  addTransports();
  addLoggers();
};

```
We init express and the logger in the app.js

```javascript
const express = require('express');

const app = express();

require('./init/initExpress')(app);
require('./init/initLogger')();

module.exports = app;
```

We use the logger for instance:
```javascript
const { loggers } = require('winston');

const logger = loggers.get('MY');
const id = 'myId';
const myRequest = 'my request';
logger.log({ level: 'info', id, message: { type: 'REQUEST', myRequest } });
```

**Comment by markwolff** [comment]
Hey @aparralorenzo thank you so much for the repro! I believe this is a different issue than OP's, since app insights uses different patches for winston v2 and v3. To give some context, I add an [app insights transport](https://github.com/Microsoft/node-diagnostic-channel/blob/ea26dc35f7a27e7e00644f764e7b875cf435711d/src/diagnostic-channel-publishers/src/winston.pub.ts#L58-L89) whenever you call `createLogger` or `configure`, as well as adding it to the "root" winston object, e.g. `winston.log('this winston has the app insights transport')`.

With your repro, I believe when you set your transports, the app insights one is being overwritten here.
```js
loggers.options.transports = developmentTransports;
```
You can try adding these transports to your loggers via a call to [`.configure(...)` or when you call `createLogger(...)`](https://github.com/winstonjs/winston#logging). This way, it will preserve the app insights transport.

Alternatively, you can add our transport to your code and modify it to just call `trackTrace` on  `appInsights.defaultClient` instead.
https://github.com/Microsoft/node-diagnostic-channel/blob/ea26dc35f7a27e7e00644f764e7b875cf435711d/src/diagnostic-channel-publishers/src/winston.pub.ts#L58-L89

**Comment by aparralorenzo** [comment]
Thank you @markwolff for the response! Can you provide an example adding AppInsightsTransport to logger.options.transport, please

**Comment by markwolff** [comment]
@aparralorenzo Sure thing. Feel free to modify the call to `trackTrace` to modify how you want the message to look. Also note that this will only work on `winston 3.x`
```js
// App Insights setup
const appInsights = require('applicationinsights');
appInsights.setup('')
    .setInternalLogging(true,true)
    .start();
appInsights.defaultClient.config.maxBatchSize = 1;

// Require in Winston 3.x
var winston = require('winston');
var {createLogger, transports, loggers} = winston;

// App Insights custom Transport
class AppInsightsTransport extends winston.Transport {
    mapLevelToKind(winston, level) {
        let levelKind;
        if (winston.config.npm.levels[level] != null) {
            levelKind = "npm";
        } else if (winston.config.syslog.levels[level] != null) {
            levelKind = "syslog";
        } else {
            levelKind = "unknown";
        }
        return levelKind;
    };

    constructor(winston, client, opts) {
        super(opts);
        this.winston = winston;
        this.client = client;
    }

    log(info, callback) {
        const {message, level} = info;
        const levelKind = this.mapLevelToKind(this.winston, level);
        const meta = {};

        if (typeof Symbol["for"] === "function") {
            const metaSearch = info.meta ? [info.meta] : undefined || info[Symbol["for"]("splat")];
            if (metaSearch) {
                for (const key in metaSearch[0]) {
                    if (meta.hasOwnProperty) {
                        meta[key] = metaSearch[0][key];
                    }
                }
            }
        }

        this.client.trackTrace({message: `${level}: ${message}`, properties: meta});
        callback();
    }
}

// Create the loggers
const myLogger = createLogger({
    level: 'info'
});
loggers.add('MY', myLogger);
loggers.options.transports = [new AppInsightsTransport(winston, appInsights.defaultClient)];

const logger = loggers.get('MY');
logger.info('hello world');

```

**Comment by visiongeist** [comment]
> u requiring in winston somewhere else first in your app (e.g. in a different file)? For the auto transporting to work, App Insights needs to be the first thing you require (or at least be before everything that you want it to monitor).

Hi. Sorry for the late answer. That worked for me. 
I am investigating why it does not work in production

**Comment by markwolff** [comment]
Closing this since it seems both issues are resolved.
@visiongeist please reopen if you are still having issues in prod
@aparralorenzo Did adding the transport work out for you? Let me know!

**Comment by aparralorenzo** [comment]
@markwolff it works! thanks!

**Comment by MrPnut** [comment]
Not to revive an old thread, but I ran into this just now and it's because I was running the app in vs code.

VS code adds 'APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL' environment variable which prevents logging the traces.  If you add the following to the top of your entrypoint script, you will get winston logs in AI.

`delete process.env['APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL'];`

I did not have to make any modifications to Winston or the transports to make this work, it worked out of the box.
